### PR TITLE
Support Memo in visibility

### PR DESCRIPTION
**Pull request by vancexu**

### CODE REVIEW

File: src/main/java/com/uber/cadence/client/WorkflowOptions.java [hunk 1]
```diff
@@ -181,6 +186,12 @@ public Builder setCronSchedule(String cronSchedule) {
       return this;
     }
 
+    /** Specifies additional information in result of list workflow. */
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/client/WorkflowOptions.java]: additional non indexable information

File: src/main/java/com/uber/cadence/client/WorkflowOptions.java [hunk 2]
```diff
@@ -356,6 +379,9 @@ public String toString() {
         + ", cronSchedule='"
         + cronSchedule
         + '\''
+        + ", memo='"
+        + memo.toString()
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/client/WorkflowOptions.java]: What if memo is null?
In Java no need to call .toString when summing up a string. Just change to "memo="  + memo +

File: src/main/java/com/uber/cadence/internal/common/StartWorkflowExecutionParameters.java [hunk 3]
```diff
@@ -363,6 +374,9 @@ public String toString() {
         + ", cronSchedule='"
         + cronSchedule
         + '\''
+        + ", memo='"
+        + memo.toString()
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/common/StartWorkflowExecutionParameters.java]: ditto

File: src/main/java/com/uber/cadence/internal/sync/WorkflowStubImpl.java [hunk 4]
```diff
@@ -136,9 +138,21 @@ private StartWorkflowExecutionParameters getStartWorkflowExecutionParameters(
...
     return p;
   }
 
+  private Map<String, byte[]> convertMemoFromObjectToBytes(Map<String, Object> memoFromOption) {
+    if (memoFromOption == null) {
+      return null;
+    }
+    Map<String, byte[]> memo = new HashMap<>();
+    for (Map.Entry<String, Object> item : memoFromOption.entrySet()) {
+      memo.put(item.getKey(), dataConverter.toData(item.getValue()));
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/internal/sync/WorkflowStubImpl.java]: I would catch the exception and add information about which field of the memo failed serialization.

File: src/main/java/com/uber/cadence/workflow/ChildWorkflowOptions.java [hunk 5]
```diff
@@ -346,6 +369,9 @@ public String toString() {
         + childPolicy
         + ", cronSchedule="
         + cronSchedule
+        + ", memo='"
+        + memo.toString()
```
Comment (mfateev) [inline src/main/java/com/uber/cadence/workflow/ChildWorkflowOptions.java]: same

File: src/test/java/com/uber/cadence/workflow/WorkflowTest.java [hunk 6]
```diff
@@ -1147,10 +1147,17 @@ private void waitForProc(WorkflowExecution execution) {
...
+    memo.put("testKey", "testObject");
+    return memo;
+  }
+
   @Test
   public void testStart() {
     startWorkerFor(TestMultiargsWorkflowsImpl.class);
-    WorkflowOptions workflowOptions = newWorkflowOptionsBuilder(taskList).build();
+    WorkflowOptions workflowOptions =
+        newWorkflowOptionsBuilder(taskList).setMemo(getTestMemo()).build();
```
Comment (mfateev) [inline src/test/java/com/uber/cadence/workflow/WorkflowTest.java]: Add check (may be using the WorkflowExecutionStarted event) that the memo was actually passed to the service.

File: src/main/java/com/uber/cadence/client/WorkflowOptions.java (hunk 2 above)
Comment (vancexu) [reply src/main/java/com/uber/cadence/client/WorkflowOptions.java]: Updated, thanks

File: src/main/java/com/uber/cadence/internal/sync/WorkflowStubImpl.java (hunk 4 above)
Comment (vancexu) [reply src/main/java/com/uber/cadence/internal/sync/WorkflowStubImpl.java]: updated

File: src/test/java/com/uber/cadence/workflow/WorkflowTest.java (hunk 6 above)
Comment (vancexu) [reply src/test/java/com/uber/cadence/workflow/WorkflowTest.java]: Done

**Pull request merged**